Cabal-described/src/Distribution/Described.hs

```diff
@@ -414,10 +414,28 @@ instance Described ForeignLibType where
     describe _ = REUnion ["native-shared","native-static"]
 
 instance Described IncludeRenaming where
-    describe _ = mr <> REOpt (RESpaces <> "requires" <> RESpaces1 <> mr)
+    -- Unfortunately, we can't directly use ModuleRenaming, as
+    -- there is some ambiguity in the grammar that we can't express
```

Comment:
Is this an expressivity problem of a framework, or is there an actual ambiguity in the grammar? I might sound like a SPJ, but can you write down the BNF-grammar then (assuming there is munching lexer, so without need to worry about spaces?)

Reply:
Note, these regexes are virtually only documentation of backpack features (re: https://github.com/haskell/cabal/issues/4761).

And while `mixin` syntax definition in https://cabal.readthedocs.io/en/3.4/buildinfo-fields-reference.html is technically correct, it goes beyond what I'd consider "human understandable".

Reply:
> Is this an expressivity problem of a framework, or is there an actual ambiguity in the grammar?

Ooh, I used a bad turn of phrase here. Ambiguity isn't the right word.

First, here is the EBNF you requested:

```
IncludeRenaming ::= ModuleRenaming { "requires" ModuleRenaming }
ModuleRenaming ::=
    (* empty *)
  | "(" Renaming ")"
  | "hiding" "(" ModuleList ")"
  | "qualified" ModuleName

RenamingList ::=
    Renaming
  | Renaming "," RenamingList

Renaming ::=
    ModuleName
  | ModuleName "as" ModuleName

ModuleList ::=
    ModuleName
  | ModuleName "," ModuleList
```

OK, now to explain the problem here. The problem I am having with Distribution.Described lies *solely* in whitespace handling, as I mentioned in https://github.com/haskell/cabal/pull/7303#issuecomment-785584905 The problem is that it is only valid to omit the space before requires (RESpaces) *if* you didn't select the "qualified" production in ModuleRenaming. If you assume that everything is tokenized beforehand, the problem disappears because `M.Nrequires` will properly tokenize as a single module name, while `)requires` will tokenize as two tokens.

> Note, these regexes are virtually only documentation of backpack field syntax (re: #4761).

Yes, no excuse for this. I guess now is a good time to nail that issue :)

Reply:
@phadej Backpack docs exist and landed!